http-server-netty/src/main/java/io/micronaut/http/server/netty/body/StreamingMultiObjectBody.java

```diff
@@ -0,0 +1,272 @@
+    }
+
+    /**
+     * A subscriber that allows blocking reads from a publisher. Handles resource cleanup properly.
+     */
+    private static final class PublisherAsBlocking implements Subscriber<Object>, Closeable {
+        private final Lock lock = new ReentrantLock();
+        private final Condition newDataCondition = lock.newCondition();
+        /**
+         * Set when {@link #take()} is called before {@link #onSubscribe}. {@link #onSubscribe} will
+         * immediately request some input.
+         */
+        private boolean pendingDemand;
+        /**
+         * Pending object, this field is used to transfer from {@link #onNext} to {@link #take}.
+         */
+        private Object swap;
+        /**
+         * The upstream subscription.
+         */
+        private Subscription subscription;
+        /**
+         * Set by {@link #onComplete} and {@link #onError}.
+         */
+        private boolean done;
+        /**
+         * Set by {@link #close}. Further objects will be discarded.
+         */
+        private boolean closed;
+        /**
+         * Failure from {@link #onError}.
+         */
+        private Throwable failure;
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            boolean pendingDemand;
+            lock.lock();
+            try {
+                this.subscription = s;
+                pendingDemand = this.pendingDemand;
+            } finally {
+                lock.unlock();
+            }
+            if (pendingDemand) {
+                s.request(1);
+            }
+        }
+
+        @Override
+        public void onNext(Object o) {
+            lock.lock();
+            try {
+                if (closed) {
+                    ReferenceCountUtil.release(o);
+                    return;
+                }
+                swap = o;
+                newDataCondition.signalAll();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        public void onError(Throwable t) {
+            lock.lock();
+            try {
+                if (swap != null) {
+                    ReferenceCountUtil.release(swap);
+                    swap = null;
+                }
+                failure = t;
+                done = true;
+                newDataCondition.signalAll();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        @Override
+        public void onComplete() {
+            lock.lock();
+            try {
+                done = true;
+                newDataCondition.signalAll();
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        /**
+         * Get the next object.
+         *
+         * @return The next object, or {@code null} if the stream is done
+         */
+        @Nullable
+        public Object take() throws InterruptedException {
+            boolean demanded = false;
+            while (true) {
```

Comment:
Why is this looping? I see it as one or nothing

Reply:
I guess because of the missing subscription

Reply:
I would implement this as `boolean consumeBlocking(Consumer c)` to handle the release in the same class

Reply:
yea it's because of the subscription, and also because await() can have a spurious wakeup.

can't do release in the same method, because PublisherAsStream stores the bytebuf for a time in case the read output buffer is too small

Reply:
Ok